github/ratelimit.go

```diff
@@ -0,0 +1,219 @@
+}
+
+func (e *ErrAbuseRateLimit) Error() string {
+	return fmt.Sprintf("abuse detection mechanism triggered; %s", e.errRetryLater.Error())
+}
+
+type errRetryLater struct {
+	retryAfter time.Time
+}
+
+func (e *errRetryLater) Error() string {
+	return fmt.Sprintf("retry after %s", e.retryAfter)
+}
+
+func (e *errRetryLater) when() time.Time {
+	return e.retryAfter
+}
+
+type whener interface {
+	error
+	when() time.Time
+}
+
+/*
+An apiErrorResponse reports one or more errors caused by an API request.
+GitHub API docs: https://developer.github.com/v3/#client-errors
+*/
+type apiErrorResponse struct {
+	Message          string `json:"message,omitempty"`
+	DocumentationURL string `json:"documentation_url,omitempty"`
+	Errors           []struct {
+		Type    string `json:"type"`
+		Message string `json:"message"`
+	} `json:"errors,omitempty"`
+}
+
+func (aer *apiErrorResponse) isAbuseRateLimit() bool {
+	return strings.Contains(aer.DocumentationURL, "abuse") ||
+		strings.Contains(aer.Message, "abuse detection")
+}
+
+// asErrAbuseRateLimit will return an ErrAbuseRateLimit if the status Response
+// is 403 Forbidden and there is a valid 'Retry-After' Response header
+// It will also return an ErrAbuseRateLimit with a default RetryAfter if the
+// response.Body can be mapped as an apiErrorResponse, and its Message or Documentation
+// mentions the abuse error.
+func asErrAbuseRateLimit(resp *http.Response, logger log.Logger) *ErrAbuseRateLimit {
+	if resp.StatusCode != http.StatusForbidden {
+		return nil
+	}
+
+	retryInHeader := resp.Header.Get("Retry-After")
+	retryIn, err := strconv.Atoi(retryInHeader)
+	if err == nil {
+		return &ErrAbuseRateLimit{
+			errRetryLater{time.Now().Add(time.Duration(retryIn) * time.Second)},
+		}
+	}
+
+	errorResponse := &apiErrorResponse{}
+	err = readAPIErrorResponse(resp, errorResponse)
+	if err == nil && errorResponse.isAbuseRateLimit() {
+		logger.Warningf("error reading 'Retry-After=%s' header from the '403 Forbidden' response, using default '%s': %s",
+			retryInHeader,
+			defaultAbuseRetryAfter,
+			err,
+		)
+
+		return &ErrAbuseRateLimit{
+			errRetryLater{time.Now().Add(defaultAbuseRetryAfter)},
+		}
+	}
+
+	logger.Warningf("403 Forbidden response got, but could not be read as an Abuse Rate Limit response")
+	return nil
+}
+
+// asErrRateLimit will return an ErrRateLimit if the Response 'X-RateLimit-Remaining' header is 0,
+// and the X-RateLimit-Reset' header contains a valid reset info
+func asErrRateLimit(resp *http.Response) *ErrRateLimit {
+	rateLimitResetHeader := resp.Header.Get("X-RateLimit-Reset")
+	rateLimitReset, errReadingReset := strconv.Atoi(rateLimitResetHeader)
+	rateLimitRemainingHeader := resp.Header.Get("X-RateLimit-Remaining")
+	rateLimitRemaining, errReadingRemaining := strconv.Atoi(rateLimitRemainingHeader)
+
+	if errReadingReset == nil && errReadingRemaining == nil && rateLimitRemaining == 0 {
+		return &ErrRateLimit{
+			errRetryLater{time.Unix(int64(rateLimitReset)+1, 0)},
```

Comment:
nit: is `+1` required?

Reply:
well... I had problems with Durations because of half seconds... and since I didn't want to deal with rounding seconds, I just added an extra second of wait, and done.

If it's not a big deal, I'd let it go.

Reply:
279e92d20d5acde364ee0a435b92c378f59c7d86 done